### PR TITLE
GameOptions page display & editing

### PR DESCRIPTION
**Pull request by TayouVR**
This reenables the existing GameOptions page and adds support for editing fields. For that it determines the type of the data in any given options. 
There is int, float, string, bool, keyBind (and technically array). Based on that it chooses the input widget used.

All currently known options and their default values are stored and will be displayed to the user.
Any options, that aren't known will just have empty fields. 
This is a Draft and the UI is going to change. In particular; I plan to implement a button to reset options to the defaultValue right into the value column for all known options.
Also sliders need the current value, min & max displayed somewhere.

TODO:
- [x] Enable GameOptions Page
- [x] Implement Editing in Model
- [x] Add Known Options Data & Columns
- [ ] Implement custom widgets for input of known options
	- [x] implement reset to default button (remove "Default Value" column in favor of tooltip on the button)
	- [x] slider widget
		- [x] display min value
		- [x] display max value
		- [x] display current value
		- [ ] allow custom unit (e.g. for FOV)
	- [x] ComboBox widget
	- [x] Checkbox Widget
	- [x] Spinner (number input)
	- [x] Text Input
	- [ ] Key Binding
		- [x] Keyboard Binding (doesn't map to minecrafts bindings one to one, modded keybinds might not even use GLFW and the same keycodes)
		- [ ] Mouse Binding
	- [ ] Bug fixing and testing
		- [x] rendering issue: the custom widget renders on top of the existing data, which is problematic with some themes like fluent, as you can see both the text in the widget and the text below
		- [ ] widget height is not enforced, breaks custom slider widget on almost all themes
- [ ] Implement Saving
- [ ] (Optional) Saving default options.txt file
	- [ ] "Set as default" button on widgets
	- [ ] Page in global settings to view (and alter) default settings (including adding custom keys)

### CODE REVIEW

File: launcher/minecraft/gameoptions/GameOptionsSchema.h [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    typename T min, max;
+};
+
+union UniversalRange {
+    Range<float> floatRange;
+    Range<int> intRange;
+};
+
+union OptionValue {
+    float floatValue;
```
Comment (Ryex) [inline launcher/minecraft/gameoptions/GameOptionsSchema.h:38]: Is a QVariant not suitable for this? it can handle all these and more. even the case of a list of variants.

**Comment by DioEgizio** [comment]
Any progress on this?

**Comment by TayouVR** [comment]
I've not had much time as I've been applying for jobs a ton the past weeks and lost a lot of motivation. I will try to work on it again within this week...

**Comment by TayouVR** [comment]
I've really been lacking motivation and I don't really know how to continue. My current progress is here (as I wanted to keep the history in the PR branch clean) https://github.com/TayouVR/PrismLauncher/tree/gameOptionsPage-wip
if anyone would like to help I'd be happy to accept any help on this, otherwise I will ofc still try to get things done.

I might also limit the scope and get rid of the slider and dropdown inputs and just leave it at SpinBox, TextBox and Checkbox for now, as implementing that doesn't seem to be as straightforward as I thought originally...
Also the KeySequence widget is difficult to work with and doesn't accept mouse inputs, so I would like to also not implement that maybe and leave it up for a later time or someone else.

I have edited the original post to separate the OptionDelegate task into various subtasks to describe the difficulty with that.

EDIT:
After some talk on discord and thinking a bit about it I have got an idea how to do more complex widgets and how to add the reset button. I only need to execute this idea now.
in a nutshell, I will create a new custom widget, that will have the reset button and there will be subclasses of the various different widgets, so that I can have one with slider and all its values, another with ComboBox, etc.
not entirely sure how this will look like, but I guess I'll see about that when I get to it

**Comment by TheKodeToad** [comment]
image
Are you aware of this?


It's just maybe a little too big.

**Comment by TayouVR** [comment]
> Are you aware of this?

I am, the sizeHint stuff I was messing with didn't work. I kinda got stuck on that... 
If you like you are free to help me :P

**Comment by TayouVR** [comment]
blind push, am having problems with my local builds rn... hopefully rebasing worked fine

EDIT: welp, lots of stuff I'll have to have a look at tomorrow or some other time..

**Comment by TayouVR** [comment]
to give a status on this

the scope of supporting e.g. keybinds and all kinds of things is way too big and implementation would be way too error prone.
We could have a key/value view for certain config formats, but I think we should instead provide a page similar to our logs page, where you can edit not just `options.txt`, but also other mods config files in `<minecraft dir>/config/`
I may open a new issue for that, if I ever get around to it. This one I will close now.
The code isn't entirely lost, I may still repurpose some of it, if it makes sense.
